types/rdf-validate-shacl/rdf-validate-shacl-tests.ts

```diff
@@ -1,23 +1,55 @@
-import { DatasetCore, DatasetCoreFactory, NamedNode, Term, BlankNode } from "rdf-js";
-import SHACLValidator = require("rdf-validate-shacl");
-import DataFactory = require("rdf-validate-shacl/src/data-factory");
-
-const shapes: DatasetCore = <any> {};
-const data: DatasetCore = <any> {};
-const factory: DatasetCoreFactory & DataFactory = <any> {};
-
-const validator = new SHACLValidator(shapes, { factory });
-const report = validator.validate(data);
-
-const conform: boolean = report.conforms;
-
-for (const result of report.results) {
-    const message: Term[] = result.message;
-    const path: BlankNode | NamedNode | null = result.path;
-    const focusNode: BlankNode | NamedNode | null = result.focusNode;
-    const severity: NamedNode | null = result.severity;
-    const sourceConstraintComponent: BlankNode | NamedNode | null = result.sourceConstraintComponent;
-    const sourceShape: BlankNode | NamedNode | null = result.sourceShape;
+import { DataFactory, DatasetCore, DatasetCoreFactory, Quad_Object } from 'rdf-js';
+import SHACLValidator = require('rdf-validate-shacl');
+import ValidationReport = require('rdf-validate-shacl/src/validation-report');
+import DataFactoryExt = require('rdf-ext/lib/DataFactory');
+import DatasetExt = require('rdf-ext/lib/Dataset');
+import NamedNodeExt = require('rdf-ext/lib/NamedNode');
+
+const shapes1: DatasetCore = <any> {};
+const data1: DatasetCore = <any> {};
+const factory1: DatasetCoreFactory & DataFactory = <any> {};
+const shapes2: DatasetExt = <any> {};
+const data2: DatasetExt = <any> {};
+const factory2: DataFactoryExt = <any> {};
+
+// $ExpectType SHACLValidator<DatasetCore<Quad, Quad>, DatasetCoreFactory<Quad, Quad, DatasetCore<Quad, Quad>> & DataFactory<Quad, Quad>>
+const validator1 = new SHACLValidator(shapes1, { factory: factory1 });
+
+// $ExpectType ValidationReport<DatasetCore<Quad, Quad>>
+const report1 = validator1.validate(data1);
+
+// $ExpectType boolean
+report1.conforms;
+
+for (const result of report1.results) {
+    result.message; // $ExpectType Quad_Object[]
+    result.path; // $ExpectType Quad_Object | null
+    result.focusNode; // $ExpectType Quad_Object | null
+    result.severity; // $ExpectType Quad_Object | null
+    result.sourceConstraintComponent; // $ExpectType Quad_Object | null
+    result.sourceShape; // $ExpectType Quad_Object | null
```

Comment:
TS 4.1 complains with:

```
ERROR: 26:5  expect  TypeScript@4.1 expected type to be:
  Quad_Object | null
got:
  Quad | BlankNode | Literal | Variable | NamedNode<string> | null
```

(`Quad_Object` is an alias for `Quad | BlankNode | Literal | Variable | NamedNode<string>`, so it's compatible.)

I can add `// Minimum TypeScript Version: 4.4`, but seems a bit heavy-handed given it does work with 4.1?

Reply:
Learnt from #54195 how to get round this.